doc/developer/design/20250307_column_names.md

```diff
@@ -0,0 +1,163 @@
+# Precise column name tracking
+
+- Associated: https://github.com/MaterializeInc/database-issues/issues/8889
+
+Our `EXPLAIN` output typically uses column numbers. It should as much
+as possible use column names.
+
+## The Problem
+
+Our `EXPLAIN WITH (humanized_expressions)` (the default) propagates
+names upwards from tables---which means we miss many opportunities for
+column names. Consider this simple example:
+
+```sql
+CREATE TABLE t(x INT, y TEXT);
+EXPLAIN SELECT x + 1 AS succ_x, y FROM t;
+```
+
+Notice that the name `y` is kept in the _named_ column reference
+`#1{y}`, but `succ_x` is lost in the column reference `#2`.
+
+```
+Explained Query:
+  Project (#2, #1{y}) // { arity: 2 }
+    Map ((#0{x} + 1)) // { arity: 3 }
+      ReadStorage materialize.public.t // { arity: 2 }
+
+Source materialize.public.t
+
+Target cluster: quickstart
+```
+
+## Success Criteria
+
+For an MIR expression `q`, define:
+
+```
+NAMESCORE(q) = |# of named column references|
+               ------------------------------
+                  |# of column references|
+```
+
+i.e., `NAMESCORE(q)` is the percentage of column references which we
+can attach a name to.
+
+We should expect to see a serious improvement in `NAMESCORE` across
+both our tests and customer queries.
+
+## Out of Scope
+
+We should _not_ expect `NAMESCORE(q) = 1.0` for all queries `q`.
+
+In particular, window functions may significantly complicate our
+ability to keep track of names.
+
+## Solution Proposal
+
+We will need to keep track of more column name information. We propose
+doing so in a few positions:
+
+1. Keep track of column names all `ScalarExpr` constructors.
+
+2. Keep track of column names on `usize` column references in
+   `*RelationExpr`.
+
+In both cases, we will work with `Option<Rc<String>>`, being careful
```

Comment:
Any reason to prefer `Option<Rc<String>>` over `Option<Rc<str>>`? It doesn't seem we'll ever modify the name. As an alternative, we could use https://crates.io/crates/compact_str to avoid the allocation in most cases.

Reply:
@ParkMyCar Another instance where an immutable `compact_str` might be interesting!

Reply:
I've already started making some progress on that, maybe I can wrap it up soon 👀

Reply:
There's a small problem with this plan I didn't anticipate: `Rc<str>` isn't `Send`, and we put `HirScalarExpr`s into the catalog. 🙃 

Is the right fix just changing things to `Arc<str>`, or are there alternatives I should consider?

Reply:
`Arc<str>` is the right approach!